### DISCUSSION TITLE
Vehicle cost as a constraint

### DISCUSSION THREAD

**Discussion by SruthyJohn** [google/or-tools#3034] [Routing (and legacy CP) questions]
I am having say 5 vehicles with various cost for travel like (100,200,75,150,210) and with different capacity. So while creating routes i need to minimize the vehicle cost involved too. So from the above sample set need to allocate vehicle3 first since it is having minimum(75) cost.
Can we achieve this with or tools?

Sample gist:
https://gist.github.com/SruthyJohn/bd6999a9eda837f81c096fc940fbc69e

**Comment by Mizux** [comment]
First you can have a vehicle fixed cost
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/ortools/constraint_solver/routing.h#L1043-L1047
then you should have one transit function per vehicles type
e.g. 
https://gist.github.com/Mizux/6d9283388150020df7542adf63577786

**Reply by SruthyJohn** [reply]
Why do we need separate transit function for each vehicle? 
TransitIndex will be same for all routes, in this case am i right? Any samples available?

**Reply by SruthyJohn** [reply]
```cs
for (int i = 0; i < data.VehicleNumber; i++)
{
    int transitIndex = routing.RegisterTransitCallback((long fromIndex, long toIndex) =>
    {
        var fromNode = manager.IndexToNode(fromIndex);
        var toNode = manager.IndexToNode(toIndex);
        return (data.DistanceMatrix[fromNode, toNode] * vehicles[i].CostPerKm);
    });
    routing.SetArcCostEvaluatorOfVehicle(transitIndex, i);
}
```

Is this the correct way to add SetArcCostEvaluatorOfVehicle for each vehicle?

**Reply by Mizux** [reply]
> Why do we need separate transit function for each vehicle?

You say  `with various cost for travel`, so I supposed each vehicles has a different travel cost i.e. not the same transit callback

> TransitIndex will be same for all routes, in this case am i right? Any samples available?

see the gist...

**Reply by Mizux** [reply]
> ```cs
> vehicles[i].CostPerKm
> ```

should be ok unless in C# the closure capture `i` by reference aka all transit callbacks will use the final value of `i`.
note: in Python this is our top 1 issue `-_-`, in C# I don't know
